feature: BridgeRouter deploy script

anna-carroll opened the pull request:
Code to Deploy the `BridgeRouter` on multiple chains. 

- intake array of chain configs & path to a core Optics system deploy
- deploy the BridgeRouter for each chain
- enroll the peer BridgeRouters with each other
- transfer ownership the BridgeRouter on each chain to Optics governance. 
- output the Bridge deploy information to a subdirectory of the core system deploy

prestwich commented:
conceptual question: 
is the bridge a core xapp that should always be deployed? if so, how should it be integrated with governance?

prestwich commented:
need to enroll the bridge routers with eachother

anna-carroll commented:
> **Discussion**: should we also add a separate deploy process for the `BridgeRouter` only, so that we don't have to re-deploy the whole suite as we iterate on the `BridgeRouter`?

**Conclusion**: we want a separate deploy process for the `BridgeRouter`. the process will intake the timestamp of the latest core system deploy, pull the networks included in that deploy from the config folder, deploy the `BridgeRouter` for each chain, enroll the peer `BridgeRouters` with each other, then pass control of the `BridgeRouter` to governance. the deploy script will output the Bridge deploy information to a `bridge` subdirectory of the existing deploy config folder

anna-carroll commented:
⚠️ **Not ready for review yet** ⚠️ 

**TODO:** 
- [ ] debug `npm run deploy-core`
- [ ] add domain for BridgeRouter enroll - options: (1) parse from config (2) query on-chain (3) add to Chain type
- [ ] output bridge deploy to subfolder of optics core deploy
- [ ] switch envy for dotenv?
- [ ] add typescript/.env to README

anna-carroll commented:
**Deprioritized**
- _xApp contract code is not in core directory, so can't verify those contracts from the same core script_  
   - possible fixes: 
      - migrate verify to typescript directory & run verify script there
      - duplicate verify script within optics-xapps directory (bad)
      - move optics-xapps contracts to optics-core (bad)
   - best solution = use typescript script once we migrate to typescript
   - don't want to block this PR on this issue 
   - moved feature to separate issue (#487) and PR (#488); removed code from this PR: https://github.com/celo-org/optics-monorepo/pull/478/commits/503383426c509f2919580b5f952b662ee47825d2

ErinHales commented:
> **Deprioritized**
> 
> * _xApp contract artifacts are in optics-xapps directory, but script is running using ethers in optics-core directory, so verifying doesn't work yet_
>   
>   * possible fixes:
>     
>     * migrate verify to typescript directory & run verify script there
>     * duplicate verify script within optics-xapps directory (bad)
>     * move optics-xapps contracts to optics-core (bad)
>   * best solution = use typescript script once we migrate to typescript
>   * don't want to block this PR on this issue
>   * moved feature to separate issue ([feature: verify bridge deploy script #487](https://github.com/celo-org/optics-monorepo/issues/487)) and PR ([Verify latest bridge deploy #488](https://github.com/celo-org/optics-monorepo/pull/488)); removed code from this PR: [5033834](https://github.com/celo-org/optics-monorepo/commit/503383426c509f2919580b5f952b662ee47825d2)

ok, yes, was going to comment on this.  I want to move the verification over to typescript.  Do you think it would be easier to merge ts migration or this first?

anna-carroll commented on typescript/optics-deploy/src/bridge.ts:
```diff
@@ -0,0 +1,236 @@
+    );
+  }
+
+  static freshFromConfig(config: ChainConfig, coreDeployPath: string): BridgeDeploy {
```
NOTE: I am working on refactoring this to have a class BaseDeploy that both Deploy and BridgeDeploy both extend, I am just trying to work out the typescript stuff between `Chain` and `OpticsChain`

prestwich replied:
ah so here's how I would lay this out

```
type Chain { ... }   // chain details (gas, name, etc)
type CoreConfig { ... }  // configuration options for the contracts to be deployed
type CoreContracts { ... } // the contracts to be deployed
type BridgeConfig { ... }  // configuration options for the contracts to be deployed
type BridgeContracts { ... } // the contracts to be deployed

class BaseDeploy { 
    readonly chain: Chain;
}

class CoreDeploy extends BaseDeploy {
    contracts: CoreContracts
    config: CoreConfig
}

class BridgeDeploy extends BaseDeploy {
    contracts: BridgeContracts
    config: BridgeConfig
}
```

prestwich replied:
fancy way

```
export type Chain { ... }
export class BaseContracts { ... }
export class BaseConfig { ... }

export class BridgeConfig extends BaseConfig { ... }
export class CoreConfig extends BaseConfig { ... }
export class BridgeContracts extends BaseContracts { ... }
export class CoreContracts extends BaseContracts { ... }

export class Deploy<T extends BaseConfig, U extends BaseContracts> { 
    readonly chain: Chain
    readonly config: T
    contracts: U
}
```

anna-carroll commented on solidity/optics-core/js/readDeployOutput.js:
```diff
@@ -0,0 +1,83 @@
+const fs = require('fs');
```
i would like to fix all of the verify-related stuff in a separate PR. there are a couple issues with it right now

anna-carroll commented:
made issues for outstanding work 
- verify script fixes #498 
- refactor `Deploy` class #499
